xdem/coreg.py

```diff
@@ -1063,3 +1081,620 @@ def coregister(reference_raster: Union[str, gu.georaster.Raster], to_be_aligned_
+        nodata_function = scipy.interpolate.RectBivariateSpline(x=north_grid, y=east_grid, z=np.isnan(aligned_dem))
+        # Initialise east and north pixel offset variables (these will be incremented up and down)
+        offset_east, offset_north, bias = 0.0, 0.0, 0.0
+
+        # Iteratively run the analysis until the maximum iterations or until the error gets low enough
+        for i in trange(self.max_iterations, disable=not verbose, desc="Iteratively correcting dataset"):
+
+            # Calculate the elevation difference and the residual (NMAD) between them.
+            elevation_difference = ref_dem - aligned_dem
+            bias = np.nanmedian(elevation_difference)
+            # Correct potential biases
+            elevation_difference -= bias
+
+            nmad = xdem.spatial_tools.nmad(elevation_difference)
+
+            assert ~np.isnan(nmad), (offset_east, offset_north)
+
+            # Stop if the NMAD is low and a few iterations have been made
+            if i > 5 and nmad < self.error_threshold:
+                if verbose:
+                    print(f"NMAD went below the error threshold of {self.error_threshold}")
+                break
+
+            # Estimate the horizontal shift from the implementation by Nuth and Kääb (2011)
+            east_diff, north_diff, _ = get_horizontal_shift(  # type: ignore
+                elevation_difference=elevation_difference,
+                slope=slope,
+                aspect=aspect
+            )
+            # Increment the offsets with the overall offset
+            offset_east += east_diff
+            offset_north += north_diff
+
+            # Calculate new elevations from the offset x- and y-coordinates
+            new_elevation = elevation_function(y=east_grid + offset_east, x=north_grid - offset_north)
+
+            # Set NaNs where NaNs were in the original data
+            new_nans = nodata_function(y=east_grid + offset_east, x=north_grid - offset_north)
+            new_elevation[new_nans >= 1] = np.nan
+
+            # Assign the newly calculated elevations to the aligned_dem
+            aligned_dem = new_elevation
+
+        self._meta["offset_east_px"] = offset_east
+        self._meta["offset_north_px"] = offset_north
+        self._meta["bias"] = bias
+        self._meta["resolution"] = resolution
+
+    def _apply_func(self, dem: np.ndarray, transform: rio.transform.Affine) -> np.ndarray:
+        """Apply the estimated x/y/z offsets to a DEM."""
+        bounds, resolution = _transform_to_bounds_and_res(dem, transform)
+        scaling_factor = self._meta["resolution"] / resolution
+
+        # Make index grids for the east and north dimensions
+        east_grid = np.arange(dem.shape[1]) * scaling_factor
+        north_grid = np.arange(dem.shape[0]) * scaling_factor
+
+        # Make a function to estimate the DEM (used to construct an offset DEM)
+        elevation_function = scipy.interpolate.RectBivariateSpline(x=north_grid, y=east_grid,
```

Comment:
This is redundant with the _fit_func. Should it not be a separate method?

Reply:
It is not really redundant. The elevation function is made to apply the sub-pixel shift.

It could theoretically be replaced with a future `apply_matrix()` function that we were discussing, but right now, there's no better way to apply the sub-pixel shift as far as I'm aware.

Reply:
What I meant is that the same 'elevation_function' is created in _fit_func and _apply_func. Maybe it could be written as a separate method of this class... Not a huge deal, but this makes things easier to maintain.
Note that I'm the one to blame here, as these are mostly inspired from my code!